##### Pandas é uma biblioteca muito poderosa que permite trabalhar com diversos tipos de dados (int, float, string ...). Ela é uma extensão do numpy, ou seja, uma melhoria para analisar arrays com diversos atributos.

In [21]:
import pandas as pd

In [22]:
data = pd.read_csv('../pandas/f500.csv')
data

company  rank  revenues  revenue_change  profits  \
0                           Walmart     1    485873             0.8  13643.0   
1                        State Grid     2    315199            -4.4   9571.3   
2                     Sinopec Group     3    267518            -9.1   1257.9   
3          China National Petroleum     4    262573           -12.3   1867.5   
4                      Toyota Motor     5    254694             7.7  16899.3   
..                              ...   ...       ...             ...      ...   
495  Teva Pharmaceutical Industries   496     21903            11.5    329.0   
496        New China Life Insurance   497     21796           -13.3    743.9   
497       Wm. Morrison Supermarkets   498     21741           -11.3    406.4   
498                             TUI   499     21655            -5.5   1151.7   
499                      AutoNation   500     21609             3.6    430.5   

     assets  profit_change                  ceo  \
0    198825           -7.2  C. Douglas McMillon   
1    489838           -6.2              Kou Wei   
2    310726          -65.0            Wang Yupu   
3    585619          -73.7        Zhang Jianhua   
4    437575          -12.3          Akio Toyoda   
..      ...            ...                  ...   
495   92890          -79.3    Yitzhak Peterburg   
496  100609          -45.6             Wan Feng   
497   11630           20.4       David T. Potts   
498   16247          195.5    Friedrich Joussen   
499   10060           -2.7   Michael J. Jackson   

                            industry                  sector  previous_rank  \
0              General Merchandisers               Retailing              1   
1                          Utilities                  Energy              2   
2                 Petroleum Refining                  Energy              4   
3                 Petroleum Refining                  Energy              3   
4           Motor Vehicles and Parts  Motor Vehicles & Parts              8   
..                               ...                     ...            ...   
495                  Pharmaceuticals             Health Care              0   
496  Insurance: Life, Health (stock)              Financials            427   
497             Food and Drug Stores      Food & Drug Stores            437   
498                  Travel Services       Business Services            467   
499              Specialty Retailers               Retailing              0   

     country           hq_location                       website  \
0        USA       Bentonville, AR        http://www.walmart.com   
1      China        Beijing, China        http://www.sgcc.com.cn   
2      China        Beijing, China        http://www.sinopec.com   
3      China        Beijing, China        http://www.cnpc.com.cn   
4      Japan         Toyota, Japan  http://www.toyota-global.com   
..       ...                   ...                           ...   
495   Israel  Petach Tikva, Israel      http://www.tevapharm.com   
496    China        Beijing, China   http://www.newchinalife.com   
497  Britain     Bradford, Britain      http://www.morrisons.com   
498  Germany      Hanover, Germany       http://www.tuigroup.com   
499      USA   Fort Lauderdale, FL     http://www.autonation.com   

     years_on_global_500_list  employees  total_stockholder_equity  
0                          23    2300000                     77798  
1                          17     926067                    209456  
2                          19     713288                    106523  
3                          17    1512048                    301893  
4                          23     364445                    157210  
..                        ...        ...                       ...  
495                         1      56960                     33337  
496                         2      54378                      8507  
497                        13      77210                      5111 

In [23]:
data_type = type(data)
print(data_type)

<class 'pandas.core.frame.DataFrame'>


In [24]:
# determina as quantidade de linhas e colunas
data.shape

(500, 17)

In [25]:
# exibe as 5 primeiras linhas
data.head()

company  rank  revenues  revenue_change  profits  assets  \
0                   Walmart     1    485873             0.8  13643.0  198825   
1                State Grid     2    315199            -4.4   9571.3  489838   
2             Sinopec Group     3    267518            -9.1   1257.9  310726   
3  China National Petroleum     4    262573           -12.3   1867.5  585619   
4              Toyota Motor     5    254694             7.7  16899.3  437575   

   profit_change                  ceo                  industry  \
0           -7.2  C. Douglas McMillon     General Merchandisers   
1           -6.2              Kou Wei                 Utilities   
2          -65.0            Wang Yupu        Petroleum Refining   
3          -73.7        Zhang Jianhua        Petroleum Refining   
4          -12.3          Akio Toyoda  Motor Vehicles and Parts   

                   sector  previous_rank country      hq_location  \
0               Retailing              1     USA  Bentonville, AR   
1                  Energy              2   China   Beijing, China   
2                  Energy              4   China   Beijing, China   
3                  Energy              3   China   Beijing, China   
4  Motor Vehicles & Parts              8   Japan    Toyota, Japan   

                        website  years_on_global_500_list  employees  \
0        http://www.walmart.com                        23    2300000   
1        http://www.sgcc.com.cn                        17     926067   
2        http://www.sinopec.com                        19     713288   
3        http://www.cnpc.com.cn                        17    1512048   
4  http://www.toyota-global.com                        23     364445   

   total_stockholder_equity  
0                     77798  
1                    209456  
2                    106523  
3                    301893  
4                    157210

In [26]:
# exibe as últimas linhas
data.tail()

company  rank  revenues  revenue_change  profits  \
495  Teva Pharmaceutical Industries   496     21903            11.5    329.0   
496        New China Life Insurance   497     21796           -13.3    743.9   
497       Wm. Morrison Supermarkets   498     21741           -11.3    406.4   
498                             TUI   499     21655            -5.5   1151.7   
499                      AutoNation   500     21609             3.6    430.5   

     assets  profit_change                 ceo  \
495   92890          -79.3   Yitzhak Peterburg   
496  100609          -45.6            Wan Feng   
497   11630           20.4      David T. Potts   
498   16247          195.5   Friedrich Joussen   
499   10060           -2.7  Michael J. Jackson   

                            industry              sector  previous_rank  \
495                  Pharmaceuticals         Health Care              0   
496  Insurance: Life, Health (stock)          Financials            427   
497             Food and Drug Stores  Food & Drug Stores            437   
498                  Travel Services   Business Services            467   
499              Specialty Retailers           Retailing              0   

     country           hq_location                      website  \
495   Israel  Petach Tikva, Israel     http://www.tevapharm.com   
496    China        Beijing, China  http://www.newchinalife.com   
497  Britain     Bradford, Britain     http://www.morrisons.com   
498  Germany      Hanover, Germany      http://www.tuigroup.com   
499      USA   Fort Lauderdale, FL    http://www.autonation.com   

     years_on_global_500_list  employees  total_stockholder_equity  
495                         1      56960                     33337  
496                         2      54378                      8507  
497                        13      77210                      5111  
498                        23      66779                      3006  
499                        12      26000                      2310

In [27]:
# podemos utlizar o atributo dtypes para saber o tipo
# de dados em cada coluna

print(data.dtypes)

company                      object
rank                          int64
revenues                      int64
revenue_change              float64
profits                     float64
assets                        int64
profit_change               float64
ceo                          object
industry                     object
sector                       object
previous_rank                 int64
country                      object
hq_location                  object
website                      object
years_on_global_500_list      int64
employees                     int64
total_stockholder_equity      int64
dtype: object


In [28]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 17 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   company                   500 non-null    object 
 1   rank                      500 non-null    int64  
 2   revenues                  500 non-null    int64  
 3   revenue_change            498 non-null    float64
 4   profits                   499 non-null    float64
 5   assets                    500 non-null    int64  
 6   profit_change             436 non-null    float64
 7   ceo                       500 non-null    object 
 8   industry                  500 non-null    object 
 9   sector                    500 non-null    object 
 10  previous_rank             500 non-null    int64  
 11  country                   500 non-null    object 
 12  hq_location               500 non-null    object 
 13  website                   500 non-null    object 
 14  years_on_g

In [29]:
# uma forma de localizar elementos desejados
data_selection = data.loc[:, "rank"]
print(data_selection)

0        1
1        2
2        3
3        4
4        5
      ... 
495    496
496    497
497    498
498    499
499    500
Name: rank, Length: 500, dtype: int64


In [30]:
# outra forma

data_selection_2 = data["rank"]
print(data_selection_2)

0        1
1        2
2        3
3        4
4        5
      ... 
495    496
496    497
497    498
498    499
499    500
Name: rank, Length: 500, dtype: int64


##### Sempre que nos deparamos com um objeto pandas 1D, é uma série. E quando vemos um objeto pandas 2D, é um DataFrame.

##### Sempre que estivermos fazendo uma seleção, precisaremos do primeiro conjunto de colchetes, mas o segundo conjunto só será necessário se precisarmos usar uma lista para fazer nossa seleção. Não há necessidade de uma lista para selecionar apenas um elemento

In [31]:
data_1 = data.loc[:, ['rank', 'country']]
print(data_1)

     rank  country
0       1      USA
1       2    China
2       3    China
3       4    China
4       5    Japan
..    ...      ...
495   496   Israel
496   497    China
497   498  Britain
498   499  Germany
499   500      USA

[500 rows x 2 columns]


In [32]:
data_2 = data.loc[:, 'sector']
print(data_2)

0                   Retailing
1                      Energy
2                      Energy
3                      Energy
4      Motor Vehicles & Parts
                ...          
495               Health Care
496                Financials
497        Food & Drug Stores
498         Business Services
499                 Retailing
Name: sector, Length: 500, dtype: object


In [33]:
# realizar várias seleções

data3 = data.loc[:, 'rank': 'assets']
print(data3)

     rank  revenues  revenue_change  profits  assets
0       1    485873             0.8  13643.0  198825
1       2    315199            -4.4   9571.3  489838
2       3    267518            -9.1   1257.9  310726
3       4    262573           -12.3   1867.5  585619
4       5    254694             7.7  16899.3  437575
..    ...       ...             ...      ...     ...
495   496     21903            11.5    329.0   92890
496   497     21796           -13.3    743.9  100609
497   498     21741           -11.3    406.4   11630
498   499     21655            -5.5   1151.7   16247
499   500     21609             3.6    430.5   10060

[500 rows x 5 columns]


In [34]:
# selecionar apenas linhas

data4 = data.loc[0]
print(data4)
print(type(data4))

company                                    Walmart
rank                                             1
revenues                                    485873
revenue_change                                 0.8
profits                                    13643.0
assets                                      198825
profit_change                                 -7.2
ceo                            C. Douglas McMillon
industry                     General Merchandisers
sector                                   Retailing
previous_rank                                    1
country                                        USA
hq_location                        Bentonville, AR
website                     http://www.walmart.com
years_on_global_500_list                        23
employees                                  2300000
total_stockholder_equity                     77798
Name: 0, dtype: object
<class 'pandas.core.series.Series'>


In [35]:
data5 = data.loc[0:2, 'rank':'country']
print(data5)
print(type(data5))

   rank  revenues  revenue_change  profits  assets  profit_change  \
0     1    485873             0.8  13643.0  198825           -7.2   
1     2    315199            -4.4   9571.3  489838           -6.2   
2     3    267518            -9.1   1257.9  310726          -65.0   

                   ceo               industry     sector  previous_rank  \
0  C. Douglas McMillon  General Merchandisers  Retailing              1   
1              Kou Wei              Utilities     Energy              2   
2            Wang Yupu     Petroleum Refining     Energy              4   

  country  
0     USA  
1   China  
2   China  
<class 'pandas.core.frame.DataFrame'>


In [36]:
'''

Selecione por rótulo	Sintaxe Explícita	            Taquigrafia comum
Fila unica	             df.loc["linha1"]	               Nenhum
Lista de linhas	        df.loc[["linha1", "linha5"]]	    Nenhum
Fatia de linhas	       df.loc["linha1":"linha5"]	       df["linha1":"linha5"]

'''

'\n\nSelecione por rótulo\tSintaxe Explícita\t            Taquigrafia comum\nFila unica\t             df.loc["linha1"]\t               Nenhum\nLista de linhas\t        df.loc[["linha1", "linha5"]]\t    Nenhum\nFatia de linhas\t       df.loc["linha1":"linha5"]\t       df["linha1":"linha5"]\n\n'

In [37]:
#pega cada valor não nulo exclusivo em uma coluna e conta quantas vezes ele aparece como 
# um valor nessa coluna
data6 = data['country'].value_counts()
print(data6)
print(type(data6))

USA             132
China           109
Japan            51
Germany          29
France           29
Britain          24
South Korea      15
Netherlands      14
Switzerland      14
Canada           11
Spain             9
Australia         7
Brazil            7
India             7
Italy             7
Taiwan            6
Russia            4
Ireland           4
Singapore         3
Sweden            3
Mexico            2
Malaysia          1
Thailand          1
Belgium           1
Norway            1
Luxembourg        1
Indonesia         1
Denmark           1
Saudi Arabia      1
Finland           1
Venezuela         1
Turkey            1
U.A.E             1
Israel            1
Name: country, dtype: int64
<class 'pandas.core.series.Series'>


In [38]:
# vamos analisar para um DataFrame

data7 = data[['sector' ,'country']]
print(data7)
print('')
print('')

print(type(data7))
print('')
print('')

print(data7.value_counts())

                     sector  country
0                 Retailing      USA
1                    Energy    China
2                    Energy    China
3                    Energy    China
4    Motor Vehicles & Parts    Japan
..                      ...      ...
495             Health Care   Israel
496              Financials    China
497      Food & Drug Stores  Britain
498       Business Services  Germany
499               Retailing      USA

[500 rows x 2 columns]


<class 'pandas.core.frame.DataFrame'>


sector               country    
Financials           USA            27
                     China          25
Energy               China          22
Health Care          USA            15
Technology           USA            14
                                    ..
Financials           India           1
                     Russia          1
Industrials          Switzerland     1
                     South Korea     1
Aerospace & Defense  Britain         1
Length: 158, dtype: int64


In [39]:
# seleciono uma coluna específica para analaisar
paises = data['country']
print(paises)
print(type(paises))

# em seguida, eu passo a contagem de valores únicos para counts_paises
counts_paises = paises.value_counts()
print(counts_paises)

# depois eu seleciono o país deseja para saber quantas vezes ele aparece
india = counts_paises['India']
print(india)


0          USA
1        China
2        China
3        China
4        Japan
        ...   
495     Israel
496      China
497    Britain
498    Germany
499        USA
Name: country, Length: 500, dtype: object
<class 'pandas.core.series.Series'>
USA             132
China           109
Japan            51
Germany          29
France           29
Britain          24
South Korea      15
Netherlands      14
Switzerland      14
Canada           11
Spain             9
Australia         7
Brazil            7
India             7
Italy             7
Taiwan            6
Russia            4
Ireland           4
Singapore         3
Sweden            3
Mexico            2
Malaysia          1
Thailand          1
Belgium           1
Norway            1
Luxembourg        1
Indonesia         1
Denmark           1
Saudi Arabia      1
Finland           1
Venezuela         1
Turkey            1
U.A.E             1
Israel            1
Name: country, dtype: int64
7


#### Por exemplo, se quisermos selecionar uma fatia de linhas em uma lista de colunas, a sintaxe que usamos será a seguinte:

##### df . loc [ "linha1" : "linha5" , [ "col1" , "col3" , "col7" ]]


##### E se quisermos selecionar uma única linha em uma fatia de colunas, a sintaxe que usamos é a seguinte:

#### df . loc [ "linha3" , "col3" : "col7" ]

###

###

####

#### Vetorização: uma forma mais rápida de calcular valores. O pandas tmabém suporta esse tipo de procedimento

In [44]:
rank_change = data["previous_rank"] - data["rank"]

print(rank_change)

print('')

print(f'Máximo: {rank_change.max()}')


print(f'Mínimo: {rank_change.min()}')

0        0
1        0
2        1
3       -1
4        3
      ... 
495   -496
496    -70
497    -61
498    -32
499   -500
Length: 500, dtype: int64

Máximo: 226
Mínimo: -500


#### Método describe: Esse método nos informa quantos valores não nulos estão contidos na série, juntamente com a média, mínimo, máximo e outras estatísticas

In [46]:
assets = data['assets']

assets.describe()

count    5.000000e+02
mean     2.436323e+05
std      4.851937e+05
min      3.717000e+03
25%      3.658850e+04
50%      7.326150e+04
75%      1.805640e+05
max      3.473238e+06
Name: assets, dtype: float64

In [49]:
# para valores do tipo string, tem-se

paises = data['country']

print(paises)
print('')


paises.describe()

0          USA
1        China
2        China
3        China
4        Japan
        ...   
495     Israel
496      China
497    Britain
498    Germany
499        USA
Name: country, Length: 500, dtype: object



count     500
unique     34
top       USA
freq      132
Name: country, dtype: object

##### unique: Número de valores exclusivos na série. Neste caso, diz-nos que existem 34 países diferentes representados na Fortune 500.


##### top: Valor mais comum da série. Os EUA são o país que mais abriga empresas da Fortune 500.

##### freq: Frequência do valor mais comum. Exatamente 132 empresas da Fortune 500 estão sediadas nos EUA.

In [51]:
rank = data['rank']

rank_desc = rank.describe()

rank_desc

print('')

prev_rank_desc = data['previous_rank']

prev_rank_desc



0        1
1        2
2        4
3        3
4        8
      ... 
495      0
496    427
497    437
498    467
499      0
Name: previous_rank, Length: 500, dtype: int64

In [63]:
# é possível perceber que há valores com classificação 0, logo, isso está errado.
# Correto: 1 a 500
# podemos utilizar o método values.counts() para analisar os valores unicos nas colunas

print(data['previous_rank'].value_counts().loc[0])
# logo, eu tenho 33 valores iguais a zero

33


Análise de dados aprimorada : ao usar valores NaN para representar dados ausentes, o restante dos dados ainda pode ser analisado e processado, permitindo uma melhor compreensão dos padrões e relacionamentos nos dados.

Legibilidade aprimorada : conjuntos de dados que usam valores NaN para representar dados ausentes são mais legíveis, pois indica claramente a presença de valores ausentes no conjunto de dados.

Representação fácil de dados ausentes : ao usar valores NaN, os dados ausentes em um conjunto de dados podem ser facilmente representados e identificados, em vez de usar valores de espaço reservado, como 0ou -1que podem dificultar ou enganar a interpretação dos resultados dos cálculos.

In [67]:
media = data[["revenues", "profits"]].median(axis=1)
print(media.max())

249758.0


In [68]:
# encontrar valores máximo apenas para colunas numéricas


max_data = data.max(numeric_only=True)
print(max_data)

rank                            500.0
revenues                     485873.0
revenue_change                  442.3
profits                       45687.0
assets                      3473238.0
profit_change                  8909.5
previous_rank                   500.0
years_on_global_500_list         23.0
employees                   2300000.0
total_stockholder_equity     301893.0
dtype: float64


In [69]:
# para obter descrição para valores não númericos, utiliza-se include=['O']

descricao_data = data.describe(include=['O'])
descricao_data


# Series.describe()método retorna um objeto de série, 
# DataFrame.describe()método retorna um objeto de dataframe

company                  ceo                       industry  \
count       500                  500                            500   
unique      500                  500                             58   
top     Walmart  C. Douglas McMillon  Banks: Commercial and Savings   
freq          1                    1                             51   

            sector country     hq_location                 website  
count          500     500             500                     500  
unique          21      34             235                     500  
top     Financials     USA  Beijing, China  http://www.walmart.com  
freq           118     132              56                       1

In [70]:
top5_rank_revenue = data[["rank", "revenues"]].head()
print(top5_rank_revenue)

   rank  revenues
0     1    485873
1     2    315199
2     3    267518
3     4    262573
4     5    254694


In [71]:
top5_rank_revenue["revenues"] = 0
print(top5_rank_revenue)

   rank  revenues
0     1         0
1     2         0
2     3         0
3     4         0
4     5         0


In [72]:
top5_rank_revenue.loc["Sinopec Group", "revenues"] = 999
print(top5_rank_revenue)

               rank  revenues
0               1.0       0.0
1               2.0       0.0
2               3.0       0.0
3               4.0       0.0
4               5.0       0.0
Sinopec Group   NaN     999.0


In [85]:
ceo = data.loc['Dow Chemical', 'ceo'] = "Jim Fitterling"
#ceo = "Jim Fitterling"
ceo

'Jim Fitterling'

In [98]:
# Além disso, podemos utilizar indexação booleana para selecionar vários
# elementos

import pandas as pd

# vamos criar um DataFrame

dados_pessoas = {'nome': ['Joao', 'Maria', 'Lucas'],
                 'numero': [12, 25, 20]}

df = pd.DataFrame(dados_pessoas)
df

nome  numero
0   Joao      12
1  Maria      25
2  Lucas      20

In [99]:
num = df['numero'] == 25
num

0    False
1     True
2    False
Name: numero, dtype: bool

In [101]:
result = df[num]
result

nome  numero
1  Maria      25

In [103]:
result = df.loc[num, 'nome']
result

1    Maria
Name: nome, dtype: object

##

In [108]:
ampersand_bool = data["sector"] == "Motor Vehicles & Parts"
data.loc[ampersand_bool,"sector"] = "Motor Vehicles and Parts"
data.loc[data["sector"] == "Motor Vehicles & Parts","sector"] = "Motor Vehicles and Parts"

# dropna = False --> impede que o Series.value_counts() 
# método exclua valores nulos ao fazer seu cálculo
prev_rank_before = data["previous_rank"].value_counts(dropna=False).head(8)

prev_rank_before

0.0      33
1.0       1
296.0     1
300.0     1
334.0     1
325.0     1
340.0     1
393.0     1
Name: previous_rank, dtype: int64

O índice da série que Series.value_counts()produz agora nos mostra floats como 471.0em vez de inteiros. A razão por trás disso é que os pandas usam o tipo inteiro NumPy, que não suporta valores NaN. Os pandas herdam esse comportamento e, nos casos em que você tenta atribuir um valor NaN a uma coluna inteira, os pandas convertem silenciosamente essa coluna em um float dtype

In [109]:
data_copy = data
data_copy

company   rank  revenues  revenue_change  \
0                               Walmart    1.0  485873.0             0.8   
1                            State Grid    2.0  315199.0            -4.4   
2                         Sinopec Group    3.0  267518.0            -9.1   
3              China National Petroleum    4.0  262573.0           -12.3   
4                          Toyota Motor    5.0  254694.0             7.7   
...                                 ...    ...       ...             ...   
496            New China Life Insurance  497.0   21796.0           -13.3   
497           Wm. Morrison Supermarkets  498.0   21741.0           -11.3   
498                                 TUI  499.0   21655.0            -5.5   
499                          AutoNation  500.0   21609.0             3.6   
Dow Chemical                        NaN    NaN       NaN             NaN   

              profits    assets  profit_change                  ceo  \
0             13643.0  198825.0           -7.2  C. Douglas McMillon   
1              9571.3  489838.0           -6.2              Kou Wei   
2              1257.9  310726.0          -65.0            Wang Yupu   
3              1867.5  585619.0          -73.7        Zhang Jianhua   
4             16899.3  437575.0          -12.3          Akio Toyoda   
...               ...       ...            ...                  ...   
496             743.9  100609.0          -45.6             Wan Feng   
497             406.4   11630.0           20.4       David T. Potts   
498            1151.7   16247.0          195.5    Friedrich Joussen   
499             430.5   10060.0           -2.7   Michael J. Jackson   
Dow Chemical      NaN       NaN            NaN       Jim Fitterling   

                                     industry                    sector  \
0                       General Merchandisers                 Retailing   
1                                   Utilities                    Energy   
2                          Petroleum Refining                    Energy   
3                          Petroleum Refining                    Energy   
4                    Motor Vehicles and Parts  Motor Vehicles and Parts   
...                                       ...                       ...   
496           Insurance: Life, Health (stock)                Financials   
497                      Food and Drug Stores        Food & Drug Stores   
498                           Travel Services         Business Services   
499                       Specialty Retailers                 Retailing   
Dow Chemical                              NaN                       NaN   

              previous_rank  country          hq_location  \
0                       1.0      USA      Bentonville, AR   
1                       2.0    China       Beijing, China   
2                       4.0    China       Beijing, China   
3                       3.0    China       Beijing, China   
4                       8.0    Japan        Toyota, Japan   
...                     ...      ...                  ...   
496                   427.0    China       Beijing, China   
497                   437.0  Britain    Bradford, Britain   
498                   467.0  Germany     Hanover, Germany   
499                     0.0      USA  Fort Lauderdale, FL   
Dow Chemical            NaN      NaN                  NaN   

                                   website  years_on_global_500_list  \
0                   http://www.walmart.com                      23.0   
1                   http://www.sgcc.com.cn                      17.0   
2                   http://www.sinopec.com                      19.0   
3                   http://www.cnpc.com.cn                      17.0   
4             http://www.toyota-global.com                      23.0   
...                                    ...                       ...   
496            http://www.newchinalife.com                       2.0   
497               http://www.morrisons.com              

In [110]:
data_copy['nova_coluna'] = 0

data_copy

company   rank  revenues  revenue_change  \
0                               Walmart    1.0  485873.0             0.8   
1                            State Grid    2.0  315199.0            -4.4   
2                         Sinopec Group    3.0  267518.0            -9.1   
3              China National Petroleum    4.0  262573.0           -12.3   
4                          Toyota Motor    5.0  254694.0             7.7   
...                                 ...    ...       ...             ...   
496            New China Life Insurance  497.0   21796.0           -13.3   
497           Wm. Morrison Supermarkets  498.0   21741.0           -11.3   
498                                 TUI  499.0   21655.0            -5.5   
499                          AutoNation  500.0   21609.0             3.6   
Dow Chemical                        NaN    NaN       NaN             NaN   

              profits    assets  profit_change                  ceo  \
0             13643.0  198825.0           -7.2  C. Douglas McMillon   
1              9571.3  489838.0           -6.2              Kou Wei   
2              1257.9  310726.0          -65.0            Wang Yupu   
3              1867.5  585619.0          -73.7        Zhang Jianhua   
4             16899.3  437575.0          -12.3          Akio Toyoda   
...               ...       ...            ...                  ...   
496             743.9  100609.0          -45.6             Wan Feng   
497             406.4   11630.0           20.4       David T. Potts   
498            1151.7   16247.0          195.5    Friedrich Joussen   
499             430.5   10060.0           -2.7   Michael J. Jackson   
Dow Chemical      NaN       NaN            NaN       Jim Fitterling   

                                     industry                    sector  \
0                       General Merchandisers                 Retailing   
1                                   Utilities                    Energy   
2                          Petroleum Refining                    Energy   
3                          Petroleum Refining                    Energy   
4                    Motor Vehicles and Parts  Motor Vehicles and Parts   
...                                       ...                       ...   
496           Insurance: Life, Health (stock)                Financials   
497                      Food and Drug Stores        Food & Drug Stores   
498                           Travel Services         Business Services   
499                       Specialty Retailers                 Retailing   
Dow Chemical                              NaN                       NaN   

              previous_rank  country          hq_location  \
0                       1.0      USA      Bentonville, AR   
1                       2.0    China       Beijing, China   
2                       4.0    China       Beijing, China   
3                       3.0    China       Beijing, China   
4                       8.0    Japan        Toyota, Japan   
...                     ...      ...                  ...   
496                   427.0    China       Beijing, China   
497                   437.0  Britain    Bradford, Britain   
498                   467.0  Germany     Hanover, Germany   
499                     0.0      USA  Fort Lauderdale, FL   
Dow Chemical            NaN      NaN                  NaN   

                                   website  years_on_global_500_list  \
0                   http://www.walmart.com                      23.0   
1                   http://www.sgcc.com.cn                      17.0   
2                   http://www.sinopec.com                      19.0   
3                   http://www.cnpc.com.cn                      17.0   
4             http://www.toyota-global.com                      23.0   
...                                    ...                       ...   
496            http://www.newchinalife.com                       2.0   
497               http://www.morrisons.com              

In [116]:
data_copy["nova_coluna"] = data_copy["previous_rank"] - data_copy["rank"]
rank_change_desc = data_copy["nova_coluna"]

data_copy['nova_coluna']

0                 0.0
1                 0.0
2                 1.0
3                -1.0
4                 3.0
                ...  
496             -70.0
497             -61.0
498             -32.0
499            -500.0
Dow Chemical      NaN
Name: nova_coluna, Length: 501, dtype: float64

In [117]:
data_copy

company   rank  revenues  revenue_change  \
0                               Walmart    1.0  485873.0             0.8   
1                            State Grid    2.0  315199.0            -4.4   
2                         Sinopec Group    3.0  267518.0            -9.1   
3              China National Petroleum    4.0  262573.0           -12.3   
4                          Toyota Motor    5.0  254694.0             7.7   
...                                 ...    ...       ...             ...   
496            New China Life Insurance  497.0   21796.0           -13.3   
497           Wm. Morrison Supermarkets  498.0   21741.0           -11.3   
498                                 TUI  499.0   21655.0            -5.5   
499                          AutoNation  500.0   21609.0             3.6   
Dow Chemical                        NaN    NaN       NaN             NaN   

              profits    assets  profit_change                  ceo  \
0             13643.0  198825.0           -7.2  C. Douglas McMillon   
1              9571.3  489838.0           -6.2              Kou Wei   
2              1257.9  310726.0          -65.0            Wang Yupu   
3              1867.5  585619.0          -73.7        Zhang Jianhua   
4             16899.3  437575.0          -12.3          Akio Toyoda   
...               ...       ...            ...                  ...   
496             743.9  100609.0          -45.6             Wan Feng   
497             406.4   11630.0           20.4       David T. Potts   
498            1151.7   16247.0          195.5    Friedrich Joussen   
499             430.5   10060.0           -2.7   Michael J. Jackson   
Dow Chemical      NaN       NaN            NaN       Jim Fitterling   

                                     industry                    sector  \
0                       General Merchandisers                 Retailing   
1                                   Utilities                    Energy   
2                          Petroleum Refining                    Energy   
3                          Petroleum Refining                    Energy   
4                    Motor Vehicles and Parts  Motor Vehicles and Parts   
...                                       ...                       ...   
496           Insurance: Life, Health (stock)                Financials   
497                      Food and Drug Stores        Food & Drug Stores   
498                           Travel Services         Business Services   
499                       Specialty Retailers                 Retailing   
Dow Chemical                              NaN                       NaN   

              previous_rank  country          hq_location  \
0                       1.0      USA      Bentonville, AR   
1                       2.0    China       Beijing, China   
2                       4.0    China       Beijing, China   
3                       3.0    China       Beijing, China   
4                       8.0    Japan        Toyota, Japan   
...                     ...      ...                  ...   
496                   427.0    China       Beijing, China   
497                   437.0  Britain    Bradford, Britain   
498                   467.0  Germany     Hanover, Germany   
499                     0.0      USA  Fort Lauderdale, FL   
Dow Chemical            NaN      NaN                  NaN   

                                   website  years_on_global_500_list  \
0                   http://www.walmart.com                      23.0   
1                   http://www.sgcc.com.cn                      17.0   
2                   http://www.sinopec.com                      19.0   
3                   http://www.cnpc.com.cn                      17.0   
4             http://www.toyota-global.com                      23.0   
...                                    ...                       ...   
496            http://www.newchinalife.com                       2.0   
497               http://www.morrisons.com              

In [118]:
industry_usa = data_copy.loc[data_copy["country"] == "USA", "industry"].value_counts().head(2)
sector_china = data_copy.loc[data_copy["country"] == "China", "sector"].value_counts().head(3)

In [122]:
industry_usa

Banks: Commercial and Savings               8
Insurance: Property and Casualty (Stock)    7
Name: industry, dtype: int64

In [121]:
sector_china

Financials     25
Energy         22
Wholesalers     9
Name: sector, dtype: int64